server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -61,23 +64,65 @@ public class KnnVectorQueryBuilder extends AbstractQueryBuilder<KnnVectorQueryBu
     public static final ParseField FILTER_FIELD = new ParseField("filter");
 
     @SuppressWarnings("unchecked")
-    public static final ConstructingObjectParser<KnnVectorQueryBuilder, Void> PARSER = new ConstructingObjectParser<>("knn", args -> {
-        List<Float> vector = (List<Float>) args[1];
-        final float[] vectorArray;
-        if (vector != null) {
-            vectorArray = new float[vector.size()];
-            for (int i = 0; i < vector.size(); i++) {
-                vectorArray[i] = vector.get(i);
-            }
+    public static final ConstructingObjectParser<KnnVectorQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "knn",
+        args -> new KnnVectorQueryBuilder((String) args[0], (float[]) args[1], (Integer) args[2], (Float) args[3])
+    );
+
+    private static float[] parseQueryVector(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.currentToken();
+        final float[] vector;
+        if (token == XContentParser.Token.START_ARRAY) {
+            vector = parseQueryVectorArray(parser);
+        } else if (token == XContentParser.Token.VALUE_STRING) {
+            vector = parseHexEncodedVector(parser);
+        } else if (token == XContentParser.Token.VALUE_NUMBER) {
+            vector = parseNumberVector(parser);
         } else {
-            vectorArray = null;
+            throw new ParsingException(parser.getTokenLocation(), format("Unknown type for provided value [%s]", parser.text()));
         }
-        return new KnnVectorQueryBuilder((String) args[0], vectorArray, (Integer) args[2], (Float) args[3]);
-    });
+        return vector;
+    }
+
+    private static float[] parseQueryVectorArray(XContentParser parser) throws IOException {
```

Comment:
This code is duplicated, maybe we could extract it to a common class?

Reply:
++ Did some refactoring in `DenseVectorFieldMapper` to avoid code duplication but could very well do the same here. Will update :)

Reply:
Addressed in 38335170227003447b013a5cb695d4a2018c0627